mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergObjectInspector.java

```diff
@@ -72,7 +75,20 @@ public ObjectInspector primitive(Type.PrimitiveType primitiveType) {
         primitiveTypeInfo = TypeInfoFactory.booleanTypeInfo;
         break;
       case DATE:
-        return IcebergDateObjectInspector.get();
+        // create the correct inspector based on whether we're working with Hive2 or Hive3 dependencies
+        // we need to do this because there is a breaking API change in DateObjectInspector between Hive2 and Hive3
+        if (MetastoreUtil.hive3PresentOnClasspath()) {
```

Comment:
Because the object inspector is a singleton, this could be done just once to set a static field. Then all this would need to do is return the `IcebergObjectInspector.DATE_INSPECTOR`. I think that would be better than running reflect code every invocation.

Reply:
And, translating these to use `DynMethods` would make it cleaner. I think both paths should use reflection and that's easy with `DynMethods`. You just provide multiple `impl` calls and the first one that is found will be used.

Reply:
yep, makes sense